c/misra/src/rules/RULE-21-25/InvalidMemoryOrderArgument.ql

```diff
@@ -0,0 +1,172 @@
+  MemoryOrder() { getDeclaringEnum().getName() = "memory_order" }
+
+  int getIntValue() { result = getValue().toInt() }
+}
+
+/* This is the only standardized memory order, allowed by RULE-21-25. */
+class AllowedMemoryOrder extends MemoryOrder {
+  AllowedMemoryOrder() { getName() = "memory_order_seq_cst" }
+}
+
+/* An expression referring to a memory order */
+class MemoryOrderConstantAccess extends EnumConstantAccess {
+  MemoryOrderConstantAccess() { getTarget() instanceof MemoryOrder }
+
+  predicate isAllowedOrder() { getTarget() instanceof AllowedMemoryOrder }
+}
+
+/* An expression with a constant value that equals a `MemoryOrder` constant */
+class MemoryOrderConstantExpr extends Expr {
+  MemoryOrder ord;
+
+  MemoryOrderConstantExpr() {
+    if this instanceof MemoryOrderConstantAccess
+    then ord = this.(MemoryOrderConstantAccess).getTarget()
+    else ord.getIntValue() = getValue().toInt()
+  }
+
+  /* Get the name of the `MemoryOrder` this expression is valued as. */
+  string getMemoryOrderString() { result = ord.toString() }
+}
+
+/**
+ * A `stdatomic.h` function which accepts a `memory_order` value as a parameter.
+ */
+class MemoryOrderedStdAtomicFunction extends Function {
+  int orderParamIdx;
+
+  MemoryOrderedStdAtomicFunction() {
+    exists(int baseParamIdx, int baseParams, string prefix, string suffix |
+      prefix = ["__", "__c11_"] and
+      suffix = ["", ".*", "_explicit"] and
+      (
+        getName().regexpMatch(prefix + ["atomic_thread_fence", "atomic_signal_fence"] + suffix) and
+        baseParamIdx = 0 and
+        baseParams = 1
+        or
+        getName()
+            .regexpMatch(prefix + ["atomic_load", "atomic_flag_clear", "atomic_flag_test_and_set"] +
+                suffix) and
+        baseParamIdx = 1 and
+        baseParams = 2
+        or
+        getName()
+            .regexpMatch(prefix + ["atomic_store", "atomic_fetch_.*", "atomic_exchange"] + suffix) and
+        baseParamIdx = 2 and
+        baseParams = 3
+        or
+        getName().regexpMatch(prefix + "atomic_compare_exchange_.*" + suffix) and
+        baseParamIdx = [3, 4] and
+        baseParams = 5
+      ) and
+      (
+        // GCC case, may have one or two inserted parameters, e.g.:
+        // __atomic_load(8, &repr->a, &desired, order)
+        // or
+        // __atomic_load_8(&repr->a, &desired, order)
+        prefix = "__" and
+        suffix = ".*" and
+        exists(int extraParams |
+          extraParams = getNumberOfParameters() - baseParams and
+          extraParams >= 0 and
+          orderParamIdx = baseParamIdx + extraParams
+        )
+        or
+        // Clang case, no inserted parameters:
+        // __c11_atomic_load(object, order)
+        suffix = "" and
+        prefix = "__c11_" and
+        orderParamIdx = baseParamIdx
+        or
+        // Non-macro case, may occur in a subset of gcc/clang functions:
+        prefix = "" and
+        suffix = "_explicit" and
```

Comment:
The `.*` suffix would also capture `_explicit`, but not be handled by this case.

Reply:
Also, prefix = "" doesn't match the above. This must have been a dead branch. Deleted.